Pre-commit: Add hook for yaml file linting

julianakwan opened the pull request:
This PR adds [`yamllint`](https://yamllint.readthedocs.io/en/latest/) as a pre-commit hook. It offers a much more comprehensive check of `yaml` files than the style checks in `test/amrex-style-files`, but is also a bit annoying in that it will pick up on our excessively long lines and the weird indenting in `gpu-build.yml` (which I have already fixed). I changed the flag for long lines from an error to a warning, as some of those lines cannot be broken without affecting readability (as it will be a link for a download or some such).

julianakwan committed:
CI: Add pre-commit hook for yamllint

Yamllint will detect incorrect formatting in yaml files:
https://yamllint.readthedocs.io/en/latest/

julianakwan committed:
CI: Configure yamllint to warn only for line-length

We have lines that are quite long in our .yml files.
Also fix identations of files.

mirenradia commented:
Thanks for this PR! I'm happy to have this as well as #123. I don't think they're mutually exclusive.

For the line length, why don't we change it to max 80 chars and also set `allow-non-breakable-words` to allow for URLs? We could set it back to "error" in this case?

julianakwan committed:
CI: Change line-length rule

This will change the maximum allowed characters to 101
and set allow-non-breakable-words and allow-non-breakable
-inline-mappings to true to allow for long web addresses.

julianakwan commented:
> Thanks for this PR! I'm happy to have this as well as #123. I don't think they're mutually exclusive.
> 
> For the line length, why don't we change it to max 80 chars and also set `allow-non-breakable-words` to allow for URLs? We could set it back to "error" in this case?

I have changed the settings for `line-length` to allow for max 100 chars and set `allow-non-breakable-words` and `allow-non-breakable-inline-mappings` to `TRUE`. 

I think the readability of some lines have improved but also some lines are now _harder_ to read as well.

mirenradia commented on .pre-commit-config.yaml:
```diff
@@ -1,10 +1,18 @@
+    rev: v1.29.0
+    hooks:
+      - id: yamllint
+        args: ['-d {rules: {line-length: {max: 101, allow-non-breakable-words: true,
+        allow-non-breakable-inline-mappings: true}}}']
```
Ironically, I think the YAML linter has added a line break here which I think breaks the intended configuration. Can you fix this?

julianakwan replied:
I don't think it's broken as I wrote a short test case: 
```
this:
  is:
    a:
      test:
        - foo: just-a-really-really-long-line-aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa
```
`yamllint` will flag this if I set: 
```
	args: ['-d {rules: {line-length: {max: 101, allow-non-breakable-words: true,
	allow-non-breakable-inline-mappings: false}}}']
```
in `.pre-commit-config.yaml` so I think it is reading the second config option. 
[The one that really annoys me](https://github.com/GRTLCollaboration/GRTeclyn/blob/develop/.github/workflows/gpu-build.yml#L66) 
```
curl -O \
  https://developer.download.nvidia.com/compute/cuda/repos/ubuntu2204/x86_64/cuda-keyring_1.1-1_all.deb
```
cannot be broken sensibly unless the link is defined separately as a environment variable. There are a couple of other places that I can refactor in `gpu-build.yml` though - let me take another look

mirenradia replied:
For legibility in the lines I've commented on, maybe it makes sense to use `>` to allow breaking the value of `args` over multiple lines but have the same indentation for those lines?

julianakwan replied:
This is not allowed: 
```
        args: >
          ['-d {rules: {line-length: {max: 101, allow-non-breakable-words: true,
          allow-non-breakable-inline-mappings: true}}}']
```
But this is allowed:
```
        args:
          ['-d {rules: {line-length: {max: 101, allow-non-breakable-words: true,
          allow-non-breakable-inline-mappings: true}}}']
```
Would you prefer that? The first one gives a python error from the `pre-commit` package.

mirenradia replied:
Yes, I think the second one is better than what is there currently.

julianakwan commented:
I think I have resolved all my issues and rebased onto `develop` again. Could you please take another look @mirenradia ? Thanks!

mirenradia reviewed (approved):
Looks good to me!
